### DISCUSSION TITLE
Oauth v2 PKCE Flow

### DISCUSSION THREAD

**Discussion by evankalaw** [tweepy/tweepy#1841] [Questions]
Hi there,

This might be a goofy question, but I'm curious - I'm trying to use the Oauth V2 PKCE flow per the documentation to get a user context and just call ```Client.get_me()```.

I can't seem to get the flow to work, and was curious, particularly about this part in the documentation from [here:](https://docs.tweepy.org/en/v4.6.0/authentication.html#oauth-2-0-authorization-code-flow-with-pkce-user-context)

```This can be used to have a user authenticate your app. Once they’ve done so, they’ll be redirected to the Callback / Redirect URI / URL you provided. You’ll need to pass that authorization response URL to fetch the access token:```

After landing in the callback, we need to provide the "authorization response url" to fetch_token for an instantiated ```Oauth2UserHandler``` to get an access token. What exactly is that?

I assumed that this is the callback url with the appropriate query string (includes state and code), but I'm getting this error:

```oauthlib.oauth2.rfc6749.errors.InvalidClientIdError: (invalid_request) Missing required parameter [code_verifier].```

Thanks for any help you can give me.

**Comment by Harmon758** [comment]
> I assumed that this is the callback url with the appropriate query string (includes state and code)

It is.

Are you using the same instance of `OAuth2UserHandler`? The code verifier should be created and set in `OAuth2UserHandler.get_authorization_url`.

**Reply by evankalaw** [reply]
Thats a great point! I was not. I'm storing that now to persist the same instance of the ```OAuth2UserHandler``` but still failing in the fetch_token call, although thankfully (?) with a different error message:

```
ERROR:__main__:Caught exception - 'OAuth2UserHandler' object has no attribute '_client'
...
...
File "path/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/tweepy/auth.py", line 222, in fetch_token
    code_verifier=self._client.code_verifier
AttributeError: 'OAuth2UserHandler' object has no attribute '_client'
```

**Reply by Harmon758** [reply]
Can you share the relevant code? `_client` should be set when the instance of `OAuth2UserHandler` is initialized.

**Reply by evankalaw** [reply]
Sure thing:

I have a route on my server where the user is landing where I'm instantiating the ```OAuth2UserHandler``` and storing that in a Flask Session to persist it when the user lands in the callback:

```
@app.route("/oauth2")
def oauth2():
    oauth2_user_handler = tweepy.OAuth2UserHandler(
        client_id="client_id_here",
        redirect_uri=callback_url,
        scope=[scopes],
        client_secret="client_secret_here",
    )

    auth_url = oauth2_user_handler.get_authorization_url()
    session["oauth2_handler"] = oauth2_user_handler

    return redirect(auth_url)
```

I'm landing in the callback after authing and grabbing the ```OAuth2UserHandler``` from the Flask Session, and attempting to grab the access token:
```
@app.route("/callback")
@use_args(
    {
        "state": fields.Str(),
        "code": fields.Str(),
    },
    location="query",
)
def twitter_callback(args):
    state = args.get("state")
    code = args.get("code")
    oauth2_user_handler = session.get("oauth2_handler")

    access_token = oauth2_user_handler.fetch_token(
        f"http://localhost:5000/callback?state={state}&code={code}"
    )

    client = tweepy.Client(access_token)
    print(client.get_me())
```

**Reply by Harmon758** [reply]
What is `session`? How is it storing the `OAuth2UserHandler` instance?

**Reply by evankalaw** [reply]
`Session` is a Flask Session which is essentially a dictionary that contains key value pairs for a particular user/session, and so I'm just storing the handler object being instantiated in the initial route to the key ```oauth2_handler``` and grabbing it from that. It saves data in a temporary directory on the server.

I'm checking the value as a sanity check inside of the callback and it seems to definitely have a Tweepy Oauth2Handler in there.

https://pythonbasics.org/flask-sessions/

**Reply by Harmon758** [reply]
I mean specifically. It's `flask.session`?

**Reply by evankalaw** [reply]
Correct, it is ```flask.session```, my bad

**Reply by Harmon758** [reply]
I believe `flask.session` is client-side and uses cookies. It's likely not able to serialize and un-serialize a complex object like `OAuth2UserHandler` properly.

**Reply by evankalaw** [reply]
I'm back!

I have a few questions for you as I've done some more digging.

- You were correct, using sessions seems to not be serializing and un-serializing the ```OAuth2UserHandler``` properly. I've found a work around for that, but I have a few more supplementary questions.

All in all, my final issues that I'm running into are as follows:

I seem to be able to generate a token properly (this one is expired, so I don't mind sharing the contents of a print statement):

```
access_token {'token_type': 'bearer', 'expires_in': 7200, 'access_token': 'd19FRE5hLXBwTHNfWnlaZ2ZKMmxLUlpBT0pUc01GYV8tQVFraDZ2aGdWeWlaOjE2NDY3ODM5ODg5NzI6MToxOmF0OjE', 'scope': ['users.read', 'tweet.read'], 'expires_at': 1646791189.0388649}
```

I assume that this means that I'm generating the token properly from the authorization response url and the ```OAuth2UserHandler```, and after further reading in the documentation / experimenting I found that because I'm using the OAuth2 PKCE Flow I need to pass "user_auth=false" into [get_me()](https://docs.tweepy.org/en/stable/client.html#tweepy.Client.get_me), e.g. ```client.get_me(user_auth=false)``` after initializing the client, because the default for get_me is to use the OAuth 1.1 user context as user_auth's default is set to true in ```get_me()``` inside of ```tweepy/tweepy/client.py```.

When I use the Auth token above, to initialize the client per the documentation, doing both ```client = tweepy.Client(access_token["access_token"])``` (this seems to be what I want to do), or ```client = tweepy.Client(access_token)```, I am still getting a 403:

```
ERROR:__main__:Caught exception - 403 Forbidden
Traceback (most recent call last):
  File "/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/webargs/core.py", line 452, in wrapper
    return func(*args, **kwargs)
  File "path", line 93, in twitter_callback_v2
    print(client.get_me(user_auth=False))
  File "/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/tweepy/client.py", line 1910, in get_me
    return self._make_request(
  File "/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/tweepy/client.py", line 118, in _make_request
    response = self.request(method, route, params=request_params,
  File "/Library/Caches/pypoetry/virtualenvs/appserv-XG89P1st-py3.9/lib/python3.9/site-packages/tweepy/client.py", line 92, in request
    raise Forbidden(response)
tweepy.errors.Forbidden: 403 Forbidden
```

Am I using the wrong thing to initialize the client I'm using to make these API calls?

I seem to be using the proper scopes in my definition of the ```OAuth2UserHandler``` as well, being ```'users.read', 'tweet.read'``` according to the [Twitter documentation for this endpoint](https://developer.twitter.com/en/docs/twitter-api/users/lookup/api-reference/get-users-me) (which was my initial concern) but they seem to be correct.

The relevant code for this section is the following 
```
@app.route("/callback")
@use_args(
    {
        "state": fields.Str(),
        "code": fields.Str(),
    },
    location="query",
)
def twitter_callback(args):
    state = args.get("state")
    code = args.get("code")

    access_token = oauth2_user_handler.fetch_token(
        f"http://localhost:5000/callback/v2?state={state}&code={code}"
    )

    client = tweepy.Client(access_token["access_token"])

    print(client.get_me(user_auth=False))
```

Thanks so much for your correspondence, I'd love your input if possible.

**Reply by Harmon758** [reply]
That should work, and I'm unable to reproduce this:
```pycon
>>> access_token = oauth2_user_handler.fetch_token("Redacted")
>>> client = tweepy.Client(access_token["access_token"])
>>> client.get_me(user_auth=False)
Response(data=<User id=1072250532645998596 name=Tweepy Testing username=TweepyDev>, includes={}, errors=[], meta={})
```
